libs/vkd3d-shader/spirv.c

```diff
@@ -4420,23 +4441,116 @@ static bool needs_private_io_variable(const struct vkd3d_shader_signature *signa
+}
+
+static uint32_t vkd3d_dxbc_compiler_get_io_variable(struct vkd3d_dxbc_compiler *compiler,
+        SpvStorageClass storage_class, uint32_t reg_idx, uint32_t array_size,
+        enum vkd3d_shader_interpolation_mode interpolation_mode, bool is_patch_constant,
+        unsigned int *out_component_count, enum vkd3d_component_type *out_component_type)
+{
+    struct vkd3d_spirv_builder *builder = &compiler->spirv_builder;
+    const struct vkd3d_shader_signature *signature;
+    unsigned int component_count, i, write_mask;
+    enum vkd3d_component_type component_type;
+    uint32_t location, var_id, *var_ids;
+
+    if (is_patch_constant)
+    {
+        signature = compiler->patch_constant_signature;
+        var_ids = compiler->patch_constant_vars;
+    }
+    else if (storage_class == SpvStorageClassInput)
+    {
+        signature = compiler->input_signature;
+        var_ids = compiler->input_vars;
+    }
+    else
+    {
+        signature = compiler->output_signature;
+        var_ids = compiler->output_vars;
+    }
+
+    component_type = VKD3D_TYPE_VOID;
+    component_count = 0;
+    write_mask = 0;
+
+    for (i = 0; i < signature->element_count; ++i)
+    {
+        const struct vkd3d_shader_signature_element *current = &signature->elements[i];
+
+        if (current->register_index != reg_idx)
+            continue;
+
+        if (current->component_type != VKD3D_TYPE_FLOAT)
+            interpolation_mode = VKD3DSIM_CONSTANT;
+
+        if (component_type == VKD3D_TYPE_VOID)
+            component_type = current->component_type;
+        else if (current->component_type != component_type)
+            component_type = VKD3D_TYPE_UINT;
+
+        write_mask |= current->mask & 0xff;
+    }
+
+    while ((1u << component_count) <= write_mask)
+        component_count++;
+
+    if (out_component_count)
+        *out_component_count = component_count;
+
+    if (out_component_type)
+        *out_component_type = component_type;
+
+    if (var_ids[reg_idx])
+        return var_ids[reg_idx];
+
+    var_id = vkd3d_dxbc_compiler_emit_array_variable(compiler, &builder->global_stream,
+        storage_class, component_type, component_count, array_size);
+
+    location = reg_idx;
+
+    if (is_patch_constant)
+    {
+        location += storage_class == SpvStorageClassInput
+            ? compiler->input_signature->element_count
+            : compiler->output_signature->element_count;
+    }
+
+    if (storage_class == SpvStorageClassOutput && is_dual_source_blending(compiler) && location < 2)
+    {
+        vkd3d_spirv_build_op_decorate1(builder, var_id, SpvDecorationLocation, 0);
+        vkd3d_spirv_build_op_decorate1(builder, var_id, SpvDecorationIndex, location);
+    }
+    else
+    {
+        vkd3d_spirv_build_op_decorate1(builder, var_id, SpvDecorationLocation, location);
+    }
+
+    if (compiler->shader_type == VKD3D_SHADER_TYPE_PIXEL && storage_class == SpvStorageClassInput)
```

Comment:
Might make more sense to decorate Patch here, considering there's an early out for `var_ids`?

Reply:
Yeah, oversight.